packages/core/src/utils/configMergerFactory.ts

```diff
@@ -1,25 +1,23 @@
-import * as _ from 'lodash';
 import { PrismConfig, PrismConfigFactory, resolveConfig } from '..';
 
+const _compact = require('lodash/compact');
+const _merge = require('lodash/merge');
+
 /**
  * Merges all passed configs. Each next config wil override each previous config.
  */
 export function configMergerFactory<C, I>(
   ...configs: Array<PrismConfig<C, I> | undefined>
 ): PrismConfigFactory<C, I> {
   return async (input: I, defaultConfig?: PrismConfig<C, I>): Promise<C> => {
-    let mergedConfigObject;
-    for (const config of configs) {
-      if (config !== undefined) {
-        mergedConfigObject = _.merge(
-          mergedConfigObject || {},
-          await resolveConfig<C, I>(input, config, defaultConfig)
-        );
-      }
-    }
-    if (mergedConfigObject === undefined) {
-      throw new Error('All configurations passed to the factory are undefined');
-    }
-    return mergedConfigObject;
+    const resolvedConfigs = await Promise.all(
+      // remove any falsy configs
+      _compact(configs)
+        // resolve each config (resolveConfig is async)
+        .map((c: C) => resolveConfig<C, I>(input, c, defaultConfig))
+    );
+
+    // merge the configs over each other, in order
+    return _merge({}, ...resolvedConfigs);
```

Comment:
@marbemac I like the refactor, looks much nicer now. I see a problem with returning empty object, though.

First problem is: {} is not instance of C (interface says we will always return C, but we return {})
This may backfire at us here:
factory.ts#37
`const configObj: Config | undefined = await configMerger(input, defaultComponents.config);`
.
.
.
```
resource = await components.router.route(
              { resources, input, config: configObj },
              defaultComponents.router
            );
```
^ what if router expected to get undefined `config`? It will never be undefined because we return `{}`. It won't be instance of C too, so even though compilation passes we may have runtime errors. Imagine `type C = { x: number }`
x is required :(
and imagine we pass {} to that something that expects C.

Reply:
Hmm, good point - how about:

```ts
import { PrismConfig, PrismConfigFactory, resolveConfig } from '..';

const _compact = require('lodash/compact');
const _merge = require('lodash/merge');

/**
 * Merges all passed configs. Each next config wil override each previous config.
 */
export function configMergerFactory<C, I>(
  ...configs: Array<PrismConfig<C, I> | undefined>
): PrismConfigFactory<C, I> {
  return async (input: I, defaultConfig?: PrismConfig<C, I>): Promise<C> => {
    const resolvedConfigs =
      // remove any falsy resolved configs
      _compact(
        await Promise.all(
          // remove falsy config props
          _compact(configs)
            // resolve each config (resolveConfig is async)
            .map((c: C) => resolveConfig<C, I>(input, c, defaultConfig))
        )
      );

    if (!resolvedConfigs.length) {
      throw new Error('All configurations passed to the factory are undefined.');
    }

    // merge the configs over each other, in order
    return _merge({}, ...resolvedConfigs);
  };
}
```